internal/tasks/account_management.go

```diff
@@ -0,0 +1,166 @@
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"slices"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/jobloop"
+	"github.com/sapcc/go-bits/logg"
+
+	"github.com/sapcc/keppel/internal/auth"
+	peerclient "github.com/sapcc/keppel/internal/client/peer"
+	"github.com/sapcc/keppel/internal/keppel"
+	"github.com/sapcc/keppel/internal/models"
+)
+
+// EnforceManagedAccounts is a job. Each task creates newly discovered accounts from the driver.
+func (j *Janitor) EnforceManagedAccounts(registerer prometheus.Registerer) jobloop.Job {
+	return (&jobloop.ProducerConsumerJob[[]string]{
+		Metadata: jobloop.JobMetadata{
+			ReadableName: "create new managed accounts",
+			CounterOpts: prometheus.CounterOpts{
+				Name: "keppel_managed_account_creations",
+				Help: "Counter for managed account creations.",
+			},
+		},
+		DiscoverTask: j.discoverManagedAccounts,
+		ProcessTask:  j.enforceManagedAccounts,
+	}).Setup(registerer)
+}
+
+func (j *Janitor) discoverManagedAccounts(_ context.Context, _ prometheus.Labels) (accountNames []string, err error) {
+	managedAccountNames, err := j.amd.ManagedAccountNames()
+	if err != nil {
+		return nil, err
+	}
+
+	var existingAccountNames []string
+	_, err = j.db.Select(&existingAccountNames, "SELECT name FROM accounts WHERE is_managed = true")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, accountName := range managedAccountNames {
+		if !slices.Contains(existingAccountNames, accountName) {
+			accountNames = append(accountNames, accountName)
+		}
+	}
+
+	var toEnforceAccountNames []string
+	_, err = j.db.Select(&toEnforceAccountNames, "SELECT name FROM accounts WHERE is_managed = true and next_account_enforcement_at < $1", j.timeNow())
+	if err != nil {
+		return nil, err
+	}
+	accountNames = append(accountNames, toEnforceAccountNames...)
+
+	if len(accountNames) == 0 {
+		return nil, sql.ErrNoRows
+	}
+
+	return accountNames, nil
+}
+
+func (j *Janitor) enforceManagedAccounts(ctx context.Context, accountNames []string, labels prometheus.Labels) error {
+	for _, accountName := range accountNames {
+		account, securityScanPolicies, err := j.amd.ConfigureAccount(models.Account{Name: accountName})
+		if err != nil {
+			return err
+		}
+
+		if account == nil {
+			accountModel, err := keppel.FindAccount(j.db, accountName)
+			if err != nil {
+				return err
+			}
+			// assume the account got already deleted
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil
+			}
+
+			_, err = j.db.Exec("UPDATE accounts SET in_maintenance = TRUE WHERE name = $1", accountName)
+			if err != nil {
+				return err
+			}
+			// avoid quering the account twice and we set this field just above via SQL
+			accountModel.InMaintenance = true
+
+			resp, err := j.processor().DeleteAccount(ctx, *accountModel)
+			if err != nil {
+				return err
+			}
+			if resp != nil {
+				logg.Error("Deleting account %s failed: %s", accountName, resp.Error)
+				return errors.New(resp.Error)
+			}
```

Comment:
I have thought about how we can best deal with the problem that account deletion is not a one-step thing, and I have a way that should work.

- When `resp` indicates to the caller that they need to do work (e.g. deleting manifests), do it right here, then recurse back into another try of `DeleteAccount`. The entire DeleteAccount thing should probably be moved into its own function to allow for easy recursion by tail call.
- When `resp` indicates to the caller that they need to wait for background tasks to happen (the whole blob deletion thing), update `NextAccountManagementAt` with a small delay (e.g. 1 minute instead of 1 hour) to come back to this account soon.